aws-sigv4/src/types.rs

```diff
@@ -24,10 +24,16 @@ pub(crate) struct CanonicalRequest {
 }
 
 impl CanonicalRequest {
-    pub(crate) fn from<B>(req: &mut Request<B>) -> Result<CanonicalRequest, Error>
+    pub(crate) fn from<B>(req: &Request<B>) -> Result<CanonicalRequest, Error>
     where
         B: AsRef<[u8]>,
     {
+        Self::from_req_payload(req, req.body())
+    }
+    pub(crate) fn from_req_payload<B>(
+        req: &Request<B>,
+        payload: impl AsRef<[u8]>,
+    ) -> Result<CanonicalRequest, Error> {
```

Comment:
I'm not sure this function carries it's weight. In what circumstances would you want to have the `payload` be separate from `B`?